### PR TITLE
IDataView to DataFrame

### PR DESCRIPTION
**Pull request by pgovind**
Implements https://github.com/dotnet/corefxlab/issues/2741
Also fixes https://github.com/dotnet/corefxlab/issues/2754

### CODE REVIEW

File: src/Microsoft.Data/StringDataFrameColumn.cs [hunk 1]
```diff
@@ -445,5 +445,24 @@ protected internal override Delegate GetDataViewGetter(DataViewRowCursor cursor)
...
             (ref ReadOnlyMemory<char> value) => value = this[cursor.Position].AsMemory();
+
+        protected internal override void AddValueUsingCursor(DataViewRowCursor cursor, DataViewSchema.Column schemaColumn)
+        {
+            long row = cursor.Position;
+            ReadOnlyMemory<char> value = default;
+            cursor.GetGetter<ReadOnlyMemory<char>>(schemaColumn)(ref value);
+            if (Length > row)
+            {
+                this[row] = new string(value.ToArray());
```
Comment (pgovind) [inline src/Microsoft.Data/StringDataFrameColumn.cs]: Not sure if there's a way to avoid this copy

File: src/Microsoft.Data/StringDataFrameColumn.BinaryOperations.cs [hunk 2]
```diff
@@ -42,54 +42,74 @@ public override DataFrameColumn Add<T>(T value, bool inPlace = false)
             return ret;
         }
 
-        public override PrimitiveDataFrameColumn<bool> ElementwiseEquals(DataFrameColumn column)
+        internal static PrimitiveDataFrameColumn<bool> ElementwiseEqualsImplementation(DataFrameColumn left, DataFrameColumn column)
```
Comment (pgovind) [inline src/Microsoft.Data/StringDataFrameColumn.BinaryOperations.cs]: Let me know if you want me to put this into its own PR. The test case I added for the IDataView->DataFrame unit test failed because this method was missing on `ArrowStringDataFrameColumn` and it was easy enough to refactor

File: tests/Microsoft.Data.DataFrame.Tests/DataFrameTests.IDataView.cs [hunk 3]
```diff
@@ -224,5 +224,19 @@ public void TestIDataViewWithNulls()
             Assert.Equal("", preview.ColumnView[14].Values[5].ToString()); // null row
             Assert.Equal("foo", preview.ColumnView[14].Values[6].ToString());
         }
+
+        [Fact]
+        public void TestDataFrameFromIDataView()
+        {
+            DataFrame df = MakeDataFrameWithAllColumnTypes(10, withNulls: false);
+            IDataView dfAsIDataView = df;
+            DataFrame newDf = DataFrame.FromIDataView(dfAsIDataView);
```
Comment (eerhardt) [inline tests/Microsoft.Data.DataFrame.Tests/DataFrameTests.IDataView.cs]: What do you think about making an extension method on `IDataView` that returns the `DataFrame`?

```C#
IDataView values = ...;
ITransformer mlModel = ...;

DataFrame df = mlModel.Transform(values).ToDataFrame();
```

vs.

```C#
IDataView values = ...;
ITransformer mlModel = ...;

DataFrame df = DataFrame.FromIDataView(mlModel.Transform(values));
```

File: src/Microsoft.Data/PrimitiveDataFrameColumn.cs [hunk 4]
```diff
@@ -652,5 +652,24 @@ protected internal override Delegate GetDataViewGetter(DataViewRowCursor cursor)
 
         private static ValueGetter<double> CreateDecimalValueGetterDelegate(DataViewRowCursor cursor, PrimitiveDataFrameColumn<decimal> column) =>
             (ref double value) => value = (double?)column[cursor.Position] ?? double.NaN;
+
+        protected internal override void AddValueUsingCursor(DataViewRowCursor cursor, DataViewSchema.Column column)
+        {
+            long row = cursor.Position;
+            T value = default;
+            cursor.GetGetter<T>(column)(ref value);
```
Comment (eerhardt) [inline src/Microsoft.Data/PrimitiveDataFrameColumn.cs]: The `Getter` delegate should be cached outside of the "loop" of cursoring over an IDataView. Getting this for each "cell" in the IDataView is going to be super slow.

Check out [https://github.com/dotnet/machinelearning/blob/master/docs/code/IDataViewImplementation.md#why-getgetter](https://github.com/dotnet/machinelearning/blob/master/docs/code/IDataViewImplementation.md#why-getgetter)

File: src/Microsoft.Data/DataFrame.IDataView.cs [hunk 5]
```diff
@@ -136,5 +136,86 @@ public override bool MoveNext()
                 return _position < _dataFrame.RowCount;
             }
         }
+
+        public static DataFrame FromIDataView(IDataView dataView)
```
Comment (eerhardt) [inline src/Microsoft.Data/DataFrame.IDataView.cs]: We should also add the ability to limit the number of rows, and which columns. See the description in the issue [https://github.com/dotnet/corefxlab/issues/2741](https://github.com/dotnet/corefxlab/issues/2741)

> And we can provide optional arguments to limit the number of rows, and to limit which columns are selected.

File: src/Microsoft.Data/StringDataFrameColumn.BinaryOperations.cs [hunk 6]
```diff
@@ -42,54 +42,74 @@ public override DataFrameColumn Add<T>(T value, bool inPlace = false)
...
+
+        internal static PrimitiveDataFrameColumn<bool> ElementwiseNotEqualsImplementation<T>(DataFrameColumn left, T value)
+        {
+            PrimitiveDataFrameColumn<bool> ret = new PrimitiveDataFrameColumn<bool>(left.Name, left.Length);
             string valString = value.ToString();
-            for (long i = 0; i < Length; i++)
+            for (long i = 0; i < left.Length; i++)
             {
-                ret[i] = (string)this[i] != valString;
+                ret[i] = (string)left[i] != valString;
```
Comment (eerhardt) [inline src/Microsoft.Data/StringDataFrameColumn.BinaryOperations.cs]: For "Arrow strings", wouldn't it be faster to convert `T value` to UTF8, and then do the comparison against the UTF8 values? That way we wouldn't need to materialize all the values as UTF16 strings.

File: src/Microsoft.Data/StringDataFrameColumn.cs (hunk 1 above)
Comment (eerhardt) [reply src/Microsoft.Data/StringDataFrameColumn.cs]: call `value.ToString()`. 

[https://source.dot.net/#System.Private.CoreLib/shared/System/ReadOnlyMemory.cs,156](https://source.dot.net/#System.Private.CoreLib/shared/System/ReadOnlyMemory.cs,156)

[https://source.dot.net/#System.Private.CoreLib/shared/System/String.Manipulation.cs,1661](https://source.dot.net/#System.Private.CoreLib/shared/System/String.Manipulation.cs,1661)

If the value is the whole backing string, it will return the whole string without copying.

---
In reply to: [338807430](https://github.com/dotnet/corefxlab/pull/2766#discussion_r338807430) [](ancestors = 338807430)

File: src/Microsoft.Data/DataFrame.IDataView.cs [hunk 7]
```diff
@@ -136,5 +136,86 @@ public override bool MoveNext()
...
+                {
+                    columns.Add(new PrimitiveDataFrameColumn<ushort>(column.Name, dataView.GetRowCount() ?? 0));
+                }
+                else if (type == TextDataViewType.Instance)
+                {
+                    columns.Add(new StringDataFrameColumn(column.Name, dataView.GetRowCount() ?? 0));
+                }
+                else
+                {
+                    throw new NotSupportedException(nameof(type));
```
Comment (eerhardt) [inline src/Microsoft.Data/DataFrame.IDataView.cs]: One issue we are going to hit immediately is that IDataView supports a "Vector" column. So in a single cell, you can store a vector of floats, for example.

We will probably need a new column type for this.

File: src/Microsoft.Data/StringDataFrameColumn.cs (hunk 1 above)
Comment (pgovind) [reply src/Microsoft.Data/StringDataFrameColumn.cs]: Interesting. Intellisense says this: `Returns the fully qualified type name of this instance` but I see that the method has different documentation on source.dot.net. I wonder why...

File: tests/Microsoft.Data.DataFrame.Tests/DataFrameTests.IDataView.cs (hunk 3 above)
Comment (pgovind) [reply tests/Microsoft.Data.DataFrame.Tests/DataFrameTests.IDataView.cs]: Yup, you brought it up before I had a chance to. I'm in agreement.

File: src/Microsoft.Data/DataFrame.IDataView.cs (hunk 7 above)
Comment (pgovind) [reply src/Microsoft.Data/DataFrame.IDataView.cs]: The rest of the feedback here is easy to fix. This one will involve some work though, since a new column has to support a bunch of functions. I'm still debating the priority here with respect to shipping a preview

Comment (eerhardt) [reply src/Microsoft.Data/DataFrame.IDataView.cs]: Feel free to log a tracking issue for this.

**Comment by pgovind** [comment]
Closing this for now. I'll reopen once we get a preview out
